gdpr_assist/models.py

```diff
@@ -246,33 +282,311 @@ def _cast_class(cls, model, privacy_meta):
+                    )
+
+                elif line["event"] == EventLog.EVENT_ALREADY_ANONYMISED:
+                    res += "{tabs}{model_name} #{target_pk} already anonymised.\n".format(
+                        tabs=tabify(indent_level),
+                        model_name=line['model_name'],
+                        target_pk=line['target_pk'],
+                    )
+
+                else:
+                    res += "{tabs}{model_name} #{target_pk} flat fields anonymised.\n".format(
+                        tabs=tabify(indent_level),
+                        model_name=line['model_name'],
+                        target_pk=line['target_pk'],
+                    )
+
+            return res
+        else:
+            return "%s is anonymised, but does not have matching logs." % self
+
+    @classmethod
+    def get_anonymisation_tree(cls, prefix="", doprint=False, objs=[]):
+        """
+        Print the result of the nesting defined above for a given model.
+        Useful for sanity and loop checking. Example output:
+
+        Class:
+        |-> outcome_assessments = (OutcomeAssessment [set_field]):
+            |-> score
+            |-> comment
+        |-> class_polls = (ClassPoll [set_field]):
+            |-> poll = (Poll [fk]):
+                |-> title
+                |-> question
+                |-> pollchoice_set = (PollChoice [set_field]):
+                    |-> choice
+                |-> pollsession_set = (PollSession [set_field]):
+                    |-> outcome_assessments = (OutcomeAssessment [set_field]):
+                        |-> score
+                        |-> comment
+                    |-> pollresponse_set = (PollResponse [set_field]):
+                        |-> response
+
+        """
+        ABANDON_AFTER_N_LEVELS = 10
+        BASE_PREFIX = "    "
+        if len(prefix) > len(BASE_PREFIX) * ABANDON_AFTER_N_LEVELS:
+            return "ERROR: shouldn't go %s levels deep, check for loops." % ABANDON_AFTER_N_LEVELS
+
+        res = ''
+        if doprint:
+            res += cls.__name__ + ":\n"
+
+        privacy_meta_model = cls._privacy_meta
+
+        flat_fields = privacy_meta_model.fields or []
+        for field in flat_fields:
+            res += "%s|-> %s\n" % (prefix, field)
+
+        for fk_field in privacy_meta_model.fk_fields:
+
+            try:
+                field = getattr(cls, fk_field)
+                if hasattr(field, 'field'):
+                    field = field.field
+                elif hasattr(field, 'rel'):
+                    field = field.rel
+                child_model = field.related_model
+                res += "%s|-> %s = (%s [fk]):\n" % (prefix, fk_field, child_model.__name__)
+
+                res += child_model.get_anonymisation_tree(prefix=BASE_PREFIX + prefix, doprint=False)
+            except AttributeError as e:
+                logger.exception("Field %s doesn't exist on %s", fk_field, cls)
+                res += 'ERROR: ' + str(e)
```

Comment:
Here again, what's the benefit of having errors in the result as opposed to failing loudly?

Reply:
see above comment.